### PR TITLE
Name GPU kernels after the current procedure and line.

### PR DESCRIPTION
**Pull request by DanilaFe**
To make it easier to figure out which GPU kernel is which, this PR makes GPU kernels be named after the procedure they're in, and the line number. It tries to skip generated procedures like `on_fn`, since most GPU kernels would be in such functions. 

Reviewed by @stonea and @e-kayrakli.

## Testing

- [x] gpu native test

### CODE REVIEW

**Review by stonea — COMMENTED** [review]
- There's the potential for naming collisions here (you could have a loop on the same line in two different files that are inside a function with the same name). Though in practice I imagine this is rare and our backend uniquifies the name.

- I think being able to specify a specific name (say with an annotation or something) would be a nice-to-have feature but no need to do that for this PR and what you have is definetly a big improvement over what we used to doo.

- It would be good to lock this behavior down with a test. I think you could just use a compopt with `--llvm-print-ir` or `--llvm-print-ir-stage`, then use a prediff that greps for the  declarations to the "gpu_kernel_" functions and a `.good` file to lock down their names.

- As demonstrated by the need to have `isParentFunction` and `isWrapperSymbol`, in order to find the
name of the function a loop was originally in takes some effort and makes some assumptions about how it may have migrated. This is a little brittle and my kneejerk reaction is to think:
  - (1) How do we know there isn't some other transformation that might migrate the loop to some other function? Is that set of wrappers complete?
    - What if the original function containing the loop is, itself, inlined?
  - (2) How can we be sure there isn't going to be some transformation in the future?

If we keep it "as is" I suppose the worst thing that can happen is a less-than ideal name (it wouldn't be a correctness issue) so I won't push hard for this.

That said, I'm tempted to say we should just use filename and line number as a more stable way of calculating this. Though doing it that way would brings up a new problem of how to sanitize the filename into something that can be an identifier.

**Comment by e-kayrakli** [comment]
> It would be good to lock this behavior down with a test. I think you could just use a compopt with --llvm-print-ir or --llvm-print-ir-stage, then use a prediff that greps for the declarations to the "gpu_kernel_" functions and a .good file to lock down their names.

If I am understanding the implementation here, we don't really need to look at the IR, do we? Like, the generated kernel names will be deterministic based on the Chapel source only. Maybe except for the numbers at the end, which can be prediffed.

(Not against this suggestion, just trying to avoid over-engineering a test)

> I'm tempted to say we should just use filename and line number as a more stable way of calculating this.

I agree with this and Andy's comments leading to this conclusion (what if the proc was `inline`?). Though, if we want to keep the procedure name in there, we can stick a field in `ForLoop` AST that records user's proc name during resolution. That can protect us against inlining and most other optimizations down the road. But then again, this might be too much over-engineering again.

---

One solution to the problem in hand is to merge this PR as is. Also add something like `--report-gpu` (not sure if it is in, yet. Still catching up.) to report kernel names. I expect this feature to be super-duper-power-user oriented, so probably should be decoupled from `--report-gpu`.

**Comment by DanilaFe** [comment]
> There's the potential for naming collisions here (you could have a loop on the same line in two different files that are inside a function with the same name). Though in practice I imagine this is rare and our backend uniquifies the name.

Right, currently we always request the name `chpl_gpu_kernel`, and that gets uniquified for us -- so here I'm relying on the same mechanism, but with extra distinguishing information. I made sure to name the kernel ".._line30" so that if it gets duplicated, it still ends up being ".._line30_2", which isn't as confusing if it does happen.

> I think being able to specify a specific name (say with an annotation or something) would be a nice-to-have feature 

It would, but I ran into some trouble getting it to work due, and thought to get this PR out first.

> It would be good to lock this behavior down with a test.

✅ 

> What if the original function containing the loop is, itself, inlined?

I was worried about it too, so I added a test. The result is a kernel name without a function name, but the correct line number -- which still seems better than the status quo.

File: compiler/optimizations/gpuTransforms.cpp [hunk 1]
```diff
@@ -661,8 +661,39 @@ GpuKernel::GpuKernel(const GpuizableLoop &gpuLoop, DefExpr* insertionPoint)
...
 }
 
+static bool isWrapperSymbol(Symbol* sym) {
+  return sym->hasFlag(FLAG_COBEGIN_OR_COFORALL_BLOCK) ||
+         sym->hasFlag(FLAG_COBEGIN_OR_COFORALL) ||
+         sym->hasFlag(FLAG_ON) ||
+         sym->hasFlag(FLAG_ON_BLOCK);
+}
+
+static Symbol* findParentFunction(Symbol* sym) {
```
Comment (stonea) [inline compiler/optimizations/gpuTransforms.cpp:671]: Add a comment indicating this function isn't reliable in the face of things like inlining.

**Review by stonea — APPROVED** [review]
Thanks for adding a test.

I'm good with this merging "as is" if you want to move on, it's a clear improvement over what we used to have.

I think I'd still prefer to have the names based on the filenames rather than function names (because of the complexity w/ `findParentFunction` and the behavior we see with inlined functions and the like).

I think it's fairly easy to get the filename from an AST node and an easy way to convert a filename into something that could be used as an identifier would be to loop over all the characters in the filename and replace any non identifier-legal character with an underscore or something.

**Review by stonea — APPROVED** [review]
Thanks for updating it to use module names!

**Pull request merged**